### PR TITLE
Added parallel iterator. From slide-rs/specs#120

### PR DESCRIPTION
**Pull request by WaDelma**
Should propably add benchmarks for it.

slide-rs/specs#120

### CODE REVIEW

File: src/iter.rs [hunk 1]
```diff
@@ -31,39 +31,194 @@ impl<T> Iterator for BitIter<T>
...
+
+#[cfg(feature="parallel")]
+pub use self::par_iter::*;
+/// A `ParallelIterator` over a [`BitSetLike`] structure.
+///
+/// [`BitSetLike`]: ../trait.BitSetLike.html
+#[cfg(feature="parallel")]
+pub struct BitParIter<T>(T);
+#[cfg(feature="parallel")]
+mod par_iter {
```
Comment (torkleyy) [inline src/iter.rs]: Would prefer this module as a file.

File: src/iter.rs [hunk 2]
```diff
@@ -31,39 +31,194 @@ impl<T> Iterator for BitIter<T>
...
+        type Item = Index;
+
+        fn drive_unindexed<C>(self, consumer: C) -> C::Result
+            where C: UnindexedConsumer<Self::Item>
+        {
+            bridge_unindexed(BitProducer((&self.0).iter()), consumer)
+        }
+    }
+
+    pub struct BitProducer<'a, T: 'a + Send + Sync>(pub BitIter<&'a T>);
```
Comment (torkleyy) [inline src/iter.rs]: Without inspecting it in detail, should this be re-exported in `lib.rs`? Or does it really stay private-only? If so, you could maybe remove `pub`.

**Review by torkleyy — COMMENTED** [review]
Great idea to have a parallel iterator for bitsets. Thank you for implementing this!

I did not take a look at the implementation details, yet.

File: src/iter.rs (hunk 1 above)
Comment (WaDelma) [reply src/iter.rs]: Hmm.. Should propably do that. Originally only reason I introduced this module was to easily feature flag it.

Comment (torkleyy) [reply src/iter.rs]: Makes sense. You can feature flag from `lib.rs`.

**Comment by torkleyy** [comment]
I guess you did not `git add` the new file. Also, we might also just have a `par_iter` file to keep the file tree flat.

**Comment by WaDelma** [comment]
Which file is missing? I don't see it. Tests seem to fail to stack overflow...

The `BitParIter` needs access to `BitIter` private fields, so it cannot be flattened (without pub(restrict)).

**Comment by torkleyy** [comment]
Now it's there! Hm, seems the recursion is too deep / does not end?

**Comment by WaDelma** [comment]
I cannot reproduce it locally and can't tell from the output which test overflowed.

File: src/iter/mod.rs [hunk 3]
```diff
@@ -1,6 +1,14 @@
 use util::*;
 use BitSetLike;
 
+#[cfg(feature="parallel")]
+pub use self::parallel::BitParIter;
```
Comment (kvark) [inline src/iter/mod.rs]: should combine as `self::parallel::{BitParIter, BitProducer};`

File: src/iter/mod.rs [hunk 4]
```diff
@@ -31,38 +39,56 @@ impl<T> Iterator for BitIter<T>
 
     fn next(&mut self) -> Option<Self::Item> {
         loop {
+            // Look at first level
```
Comment (kvark) [inline src/iter/mod.rs:41]: niice!

File: src/iter/parallel.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
+use rayon::iter::ParallelIterator;
+use rayon::iter::internal::{UnindexedProducer, UnindexedConsumer, Folder, bridge_unindexed};
+use super::*;
```
Comment (kvark) [inline src/iter/parallel.rs]: let's not do this, considered harmful in Rust style to be used anywhere outside of test modules

File: src/lib.rs [hunk 6]
```diff
@@ -15,6 +17,10 @@ mod util;
 
 pub use atomic::AtomicBitSet;
 pub use iter::BitIter;
+#[cfg(feature="parallel")]
+pub use iter::BitParIter;
```
Comment (kvark) [inline src/lib.rs]: similarly, let's combine those public use cases

File: src/lib.rs [hunk 7]
```diff
@@ -333,6 +347,27 @@ mod tests {
...
     #[test]
+    fn iter_random_add() {
+        use std::num::Wrapping as Z;
+        let mut set = BitSet::new();
+        let mut state = Z(0u64);
+        let max_added = 1_048_576 / 10;
+        let mut added = 0;
+        for _ in 0..max_added {
+            let oldstate = state;
+            state = oldstate * Z(6364136223846793005) + Z(1);
```
Comment (kvark) [inline src/lib.rs]: what is this magic? We could just use the `rand` crate for testing instead

File: src/iter/parallel.rs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+/// An `ParallelIterator` over a [`BitSetLike`] structure.
+///
+/// [`BitSetLike`]: ../../trait.BitSetLike.html
+pub struct BitParIter<T>(T);
+
+#[cfg(target_pointer_width="64")]
+pub const BITS_IN_USIZE: usize = 64;
```
Comment (kvark) [inline src/iter/parallel.rs]: this const is only used in one function, so let's just take it via `size_of` in the code - should be simpler

File: src/iter/parallel.rs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    }
+}
+
+
+/// Allows splitting and internally iterating through `BitSet`.
+///
+/// Usually used internally by `BitParIter`.
+pub struct BitProducer<'a, T: 'a + Send + Sync>(pub BitIter<&'a T>);
+
+impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
```
Comment (kvark) [inline src/iter/parallel.rs]: Being naive, I thought you can use rayon's stuff on any iterator. It's great to see you implementing this! We need to carefully inspect the math here, especially in the context of CI failures.

File: src/lib.rs (hunk 7 above)
Comment (WaDelma) [reply src/lib.rs]: Yeah, should probably do that. I didn't want to include rand at first just for tests, but if they are dev-dependency it should be fine.

(That magic is generating random numbers with [PCG](http://www.pcg-random.org/))

**Comment by WaDelma** [comment]
I do no recursion on my end, so the recursion is happening in rayon I think. I have debug it more deeply, but my first guess is that I am splitting too many times (or infinitely).

I also realized yesterday that I am using 32bit rust on my computer (I tried to install 64bit version, but trying to `rustup run cargo test` with that toolchain results into infinite recursion‽), so that's probably the difference between travis and local test.

**Comment by WaDelma** [comment]
I was able to test it with 64bit rust and couldn't reproduce stack overflow with it either...
I even tried test where I fill whole bitset with ones and it still didn't overflow.

So I have no clue what is wrong and now it goes through travis too.

**Comment by kvark** [comment]
Well, this is unsettling. I'll try to make a close look at the logic then to spot a potential error.
Edit: Also, thank you for addressing my comments! I know it's not the easiest PR ever. Given that the crate is rather fundamental and low level, we can't take much risks.

**Comment by WaDelma** [comment]
Well I don't expect this getting merged anytime soon: When I started it I had no clue how it would work and I am not 100% confident it's correct now.

File: .travis.yml [hunk 10]
```diff
@@ -9,3 +9,5 @@ script:
   - cargo doc
   - cargo test
```
Comment (kvark) [inline .travis.yml]: we need to add testing with no features enabled as well

File: src/iter/parallel.rs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    where T: BitSetLike
+{
+    type Item = Index;
+    fn split(mut self) -> (Self, Option<Self>) {
+        let other = {
+            let mut handle_level = |level: usize| {
+                // Check that the level isn't empty
+                if self.0.masks[level] != 0 {
```
Comment (kvark) [inline src/iter/parallel.rs]: let's just early-out here (with `return None`) instead of indenting the whole thing

File: src/iter/parallel.rs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                    // Find first bit that is set
+                    let first_bit = self.0.masks[level].trailing_zeros();
+                    // Find last bit that is set
+                    let last_bit = (size_of::<usize>() * 8) as u32 - self.0.masks[level].leading_zeros() - 1;
+                    // Check that there is more than one bit that is set
+                    if first_bit != last_bit {
```
Comment (kvark) [inline src/iter/parallel.rs]: again, easier to early-out here

File: src/iter/parallel.rs [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                        // Make the split point to be the avarage of first and last bit
+                        let avarage = (first_bit + last_bit) / 2;
```
Comment (kvark) [inline src/iter/parallel.rs]: typo "avarage"

File: src/iter/parallel.rs [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+                        // A bit mask to get the lower half of the mask
+                        let mask = (1 << avarage) - 1;
+                        let level_prefix = if level == self.0.prefix.len() {
+                            0
+                        } else {
+                            self.0.prefix[level]
```
Comment (kvark) [inline src/iter/parallel.rs]: alternatively, could be `*self.0.prefix.get(level).unwrap_or(&0)`

File: src/iter/parallel.rs [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+                            masks: [0, 0, 0, 0],
+                            prefix: [0, 0, 0],
+                        });
+                        // Take the higher half of the mask
+                        other.0.masks[level] = self.0.masks[level] & !mask;
+                        // The higher half starts iterating from the avarage
+                        other.0.prefix[level - 1] = (level_prefix | avarage) << BITS;
+                        // And preserves the prefix of higher levels
+                        for n in level..self.0.prefix.len() {
+                            other.0.prefix[n] = self.0.prefix[n];
```
Comment (kvark) [inline src/iter/parallel.rs]: I believe, this could be faster rewritten as:
```rust
for (mut out_prefix, in_prefix) in (other.0.prefix[level..].iter_mut(), self.0.prefix[level..].iter()) {
  *out_prefix = *in_prefix;
}
```

File: src/iter/parallel.rs [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+                        } else {
+                            self.0.prefix[level]
+                        };
+                        let mut other = BitProducer(BitIter {
+                            set: self.0.set,
+                            masks: [0, 0, 0, 0],
+                            prefix: [0, 0, 0],
+                        });
+                        // Take the higher half of the mask
+                        other.0.masks[level] = self.0.masks[level] & !mask;
```
Comment (kvark) [inline src/iter/parallel.rs]: shouldn't we copy the higher level masks to `other` as well?

File: src/lib.rs [hunk 17]
```diff
@@ -333,6 +347,22 @@ mod tests {
     }
 
     #[test]
+    fn iter_random_add() {
+        use rand::Rng;
+        let mut set = BitSet::new();
+        let mut rng = ::rand::weak_rng();
```
Comment (kvark) [inline src/lib.rs]: let's be more consistent here - since we have a `use rand::` statement on top, let's include `weak_rng` there as well.

File: src/iter/parallel.rs (hunk 16 above)
Comment (WaDelma) [reply src/iter/parallel.rs]: I am not actually sure anymore: I remember I left it out earlier for some reason. Should have written it down.

If I try to add it, it will make tests overflow their stacks...

File: src/iter/parallel.rs (hunk 15 above)
Comment (WaDelma) [reply src/iter/parallel.rs]: I could actually use `copy_from_slice` here.

File: src/iter/parallel.rs (hunk 14 above)
Comment (WaDelma) [reply src/iter/parallel.rs]: Couldn't use this code exactly, because `&0` doesn't live long enough, but used `cloned` for it instead.

File: src/iter/parallel.rs (hunk 16 above)
Comment (WaDelma) [reply src/iter/parallel.rs]: Ah... I actually did the copying wrong: I started it from current level and not one above.

I also realised why copying higher levels isn't nesessary: Each time split happens the highest level of the mask gets halved and eventually there is only one bit left there. This means that all masks of higher levels are unnessary and can be left to 0.

Comment (kvark) [reply src/iter/parallel.rs]: Awesome! Can we have a `debug_assert` expressing this precondition?

File: src/iter/parallel.rs (hunk 15 above)
Comment (kvark) [reply src/iter/parallel.rs]: oh indeed 👍

File: src/iter/parallel.rs [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+                other.0.masks[level] = self.0.masks[level] & !mask;
+                // The higher half starts iterating from the average
+                other.0.prefix[level - 1] = (level_prefix | average) << BITS;
+                // And preserves the prefix of the higher levels
+                other.0.prefix[level..].copy_from_slice(&self.0.prefix[level..]);
+                // Take the lower half the mask
+                self.0.masks[level] &= mask;
+                // The lower half starts iterating from the first bit
+                self.0.prefix[level - 1] = (level_prefix | first_bit) << BITS;
+                Some(other)
```
Comment (kvark) [inline src/iter/parallel.rs]: can we add something like `debug_assert_eq!(self + other, BitMask::all())` ?

**Comment by WaDelma** [comment]
Welp... I started to add those asserts you wanted and I actually realised that there is huge oversight: When there is only one bit left descending to the next layer doesn't actually work as the lower levels mask is never set to anything but 0. This means that it couldn't actually split more finely than for the highest level.

When I fixed that I ran into more problems. One of them is that the avarage is rounded down which meant that sometimes it thought that it could split, but splitting actually didn't change anything and there was infinite recursion. After that it still runs to problem that I haven't yet debugged.

I should really look into making the tests better. Biggest problem with them is that it uses the machinary of the normal iterator which masks these kinds of problems as the iteration is still correct, but not split enough. 

This PR and the one in specs has turned into quite an adventure :D

File: src/iter/parallel.rs (hunk 18 above)
Comment (WaDelma) [reply src/iter/parallel.rs]: I am not sure what you mean by this comment. So are the self and other their masks? If so then their sum shouldn't be equal with mask with all bits set (or do you mean something different with `BitMask::all`).

Comment (kvark) [reply src/iter/parallel.rs]: Well, conceptually this method splits the hibitset into 2 sub-sets. What I want to see is a run-time guarantee that these sub-sets don't intersect and their union equals to the original set.

Comment (WaDelma) [reply src/iter/parallel.rs]: I added check for this.. I am not sure if it's that helpful one though.

**Comment by torkleyy** [comment]
@WaDelma IIRC the logic should work now? Would you please rebase onto the latest master?

**Comment by WaDelma** [comment]
AFAIK it should work correctly now. I rebased it.

**Comment by torkleyy** [comment]
@WaDelma Great, thanks! I couldn't spot any stylistic or doc issues, and I'd rather delegate reviewing the logic to @kvark 

Really amazing work here!

**Comment by kvark** [comment]
@torkleyy I somewhat reviewed it once more. It would be best to have @csherratt glancing at the code. If they don't make it, we can merge and then fix post-release, given that the API will not change. After all, this is an extra feature, not a change in existing logic.

**Comment by csherratt** [comment]
This is seriously cool. I don't see anything I object to in the implementation.

But I would ask that we add a few more targeted tests. All of the tests will tend to fill the tree and maybe leave some empty spaces, even the random number one has no guarantee of this. It would be good to see entire leafs that were empty to make sure there isn't some ugly edge case we overlooked.

Something like this: https://play.rust-lang.org/?gist=387d26b94a6b3cddcd2eb4ff085a4d7e&version=stable&backtrace=0

That should generate clusters of keys that don't completely fill the l3/l2 nodes.

**Comment by kvark** [comment]
@csherratt thanks for expertise! I hope you'll join us on Gitter, like the olde good times ;)

**Comment by WaDelma** [comment]
Added the clustered test. Any other ideas for tests?

**Comment by kvark** [comment]
Thanks @WaDelma !
